c++/src/kj/compat/http.h

```diff
@@ -904,7 +919,7 @@ struct HttpClientSettings {
 };
 
 kj::Own<HttpClient> newHttpClient(kj::Timer& timer, const HttpHeaderTable& responseHeaderTable,
-                                  kj::Network& network, kj::Maybe<kj::Network&> tlsNetwork,
+                                  kj::Network& network, kj::Maybe<kj::TlsNetwork&> tlsNetwork,
```

Comment:
This is a backwards-incompatible API change which also narrows this function's capabilities by replacing an abstract interface with a concrete implementation. The new dependency on KJ TLS with #ifdefs that avoid it based on whether OpenSSL is available is all kind of ugly, too.

Could HttpClientSettings instead be extended with an optional callback function that could be used here, without introducing any direct dependency on `tls.h`?

Reply:
The new option could be:

`kj::Function<kj::Own<kj::AsyncIoStream>(kj::Own<kj::AsyncIoStream> inner, kj::StringPtr expectedHostname)> startTls;`

Then again, if a new option is required anyway, then perhaps it's not worth extending the default HttpClient implementation to support it. People that need this functionality might as well use a wrapper? (That's what we're going to be doing anyway.)

Reply:
Perhaps another possible approach is to not extend `HttpConnectSettings` at all, but instead pass down this callback as part of `SubrequestMetadata` in the workers runtime. Ugly in its own way but keeps the weirdness out of KJ... I dunno.

Reply:
Maybe even more simply: could `HttpClientSettings` hold a `TlsContext&`? This is the only reason I need the TlsNetwork.

Reply:
That still introduces a dependency on kj-tls. I'd like to avoid that.

What you could do is factor out an abstract interface which `TlsContext` implements. Maybe it could be called `SecureNetworkWrapper`. Put it in some common location that both kj-http and kj-tls depend on... probably `async-io.h`.